File: backend-v5/lms/forum/ForumTopic.php
<?php

// Modul: lms/forum
// Klasa: ForumTopic
// Opis: jedna tema na forumu

require_once(Config::$backend_path."core/DB.php");

// FIXME: u bazi ne postoji tabela forum
// polje forum tabele topic se zove "projekat"


class ForumTopic {
	public $id;
	public $lastUpdate, $firstPostId, $lastPostId, $views, $authorId, $forumId;
	public $lastPost;

	public static function fromId($id) {
		$q10 = DB::query("select FROM_UNIXTIME(vrijeme), prvi_post, zadnji_post, pregleda, osoba, projekat from bb_tema where id=$id");
		if (mysql_num_rows($q10)<1) {
			throw new Exception("unknown forum topic");
		}
		
		$ft = new ForumTopic;
		$ft->id = $id;
		$ft->lastUpdate = mysql_result($q10,0,0);
		$ft->firstPostId = mysql_result($q10,0,1);
		$ft->lastPostId = mysql_result($q10,0,2);
		$ft->views = mysql_result($q10,0,3);
		$ft->authorId = mysql_result($q10,0,4);
		$ft->forumId = mysql_result($q10,0,5);
		
		return $ft;
	}

	public function getCountReplies() {
		$q10 = DB::query("select count(*) from bb_post where tema=".$this->id);
		return mysql_result($q10,0,0)-1;
	}

	public function viewed() {
		$q10 = DB::query("UPDATE bb_tema SET pregleda=pregleda+1 WHERE id=".$this->id);
	}

	public function getAllPosts($limit = 0, $offset = 0) {
		if ($limit>0 && $offset>0) 
			$limitsql = "LIMIT $offset, $limit";
		else if ($limit>0)
			$limitsql = "LIMIT $limit";
		else
			$limitsql = "";

		$q10 = DB::query("
SELECT p.id, p.naslov, UNIX_TIMESTAMP(p.vrijeme), p.osoba, p.tema, pt.tekst, o.ime, o.prezime, o.brindexa
FROM bb_post as p, bb_post_text as pt, osoba as o
WHERE p.tema=".$this->id." AND p.id=pt.post AND p.osoba=o.id
ORDER BY p.vrijeme ASC
$limitsql");
		$posts = array();
		while ($r10 = mysql_fetch_row($q10)) {
			$fp = new ForumPost;
			$fp->id = $r10[0];
			$fp->subject = $r10[1];
			$fp->time = $r10[2];
			$fp->authorId = $r10[3];
			$fp->topicId = $r10[4];
			$fp->text = $r10[5];

			$fp->author = new Person;
			$fp->author->id = $r10[3];
			$fp->author->name = $r10[6];
			$fp->author->surname = $r10[7];
			$fp->author->studentIdNr = $r10[8];
			$fp->author->login = ""; // FIXME: problem je što može postojati više slogova u tabeli auth za jedan login a hoću da izbjegnem podupite

			array_push($posts, $fp);
		}
		
		return $posts;
	}

	public function addReply($forumPost) {
		$q10 = DB::query("select id from bb_post order by id desc limit 1");
		if (mysql_num_rows($q10)==0)
			$newId = 1;
		else
			$newId = mysql_result($q10,0,0)+1;

		$forumPost->id = $newId;
		$forumPost->topicId = $this->id;

		$q10 = DB::query("insert into bb_post set id=$newId, naslov='".$forumPost->subject."', vrijeme=NOW(), osoba=".$forumPost->authorId.", tema=".$this->id);

		$q20 = DB::query("select UNIX_TIMESTAMP(vrijeme) from bb_post where id=$newId");
		$forumPost->time = mysql_result($q20,0,0);

		$q30 = DB::query("insert into bb_post_text set post=$newId, tekst='".$forumPost->text."'");
		return $forumPost;
	}
}

?>

File: rorbackend/app/models/lms/forum/forum_topic.rb
class Lms::Forum::ForumTopic < ActiveRecord::Base
  # Uncomment following lines if working with legacy database
  # set_table_name 'bb_tema'
  # set_primary_key :id
  # alias_attribute :id, :id
  # alias_attribute :last_update, :vrijeme
  # alias_attribute :first_post_id, :prvi_post
  # alias_attribute :last_post_id, :zadnji_post
  # alias_attribute :views, :pregleda
  # alias_attribute :author_id, :osoba
  # alias_attribute :forum_id, :projekat

  # Uncomment following lines if working with legacy database
  # TABLE_NAME = 'bb_tema'
  # ID = TABLE_NAME + '.' + 'id'
  # LAST_UPDATE = TABLE_NAME + '.' + 'vrijeme'
  # FIRST_POST_ID = TABLE_NAME + '.' + 'prvi_post'
  # LAST_POST_ID = TABLE_NAME + '.' + 'zadnji_post'
  # VIEWS = TABLE_NAME + '.' + 'pregleda'
  # AUTHOR_ID = TABLE_NAME + '.' + 'osoba'
  # FORUM_ID = TABLE_NAME + '.' + 'projekat'

  # Comment following lines if working with legacy database
  TABLE_NAME = 'lms_forum_forum_topics'
  ID = TABLE_NAME + '.' + 'id'
  LAST_UPDATE = TABLE_NAME + '.' + 'last_update'
  FIRST_POST_ID = TABLE_NAME + '.' + 'first_post_id'
  LAST_POST_ID = TABLE_NAME + '.' + 'last_post_id'
  VIEWS = TABLE_NAME + '.' + 'views'
  AUTHOR_ID = TABLE_NAME + '.' + 'author_id'
  FORUM_ID = TABLE_NAME + '.' + 'forum_id'

  ALL_COLUMNS = [ID, LAST_UPDATE, FIRST_POST_ID, LAST_POST_ID, VIEWS, AUTHOR_ID, FORUM_ID]
  
  belongs_to :first_post, :class_name => "Lms::Forum::ForumPost"
  belongs_to :last_post, :class_name => "Lms::Forum::ForumPost"
  belongs_to :author, :class_name => "Core::Person"
  belongs_to :forum
  has_many :forum_posts
  before_save :set_time_to_now
  
  def self.get_replies_count(id)
    replies_count = (Lms::Forum::ForumPost).where(:forum_topic_id => id).count
    replies_count -= 1
    
    return replies_count
  end
  
  def self.viewed(id)
    forum_topic = (Lms::Forum::ForumTopic).find(id)
    forum_topic[:views] += 1
    return forum_topic.save
  end
  
  
  
  def self.get_all_posts(id, limit)
    limit = nil if limit == '0'
    select_columns = (Lms::Forum::ForumPost).ALL_COLUMNS | [(Core::Forum::ForumPostText).TEXT, (Core::People).ID, (Core::People).NAME, (Core::People).SURNAME, (Core::People).STUDENT_ID_NUMBER]
    topic_posts = (Lms::Forum::ForumPost).where(:forum_topic_id => id).joins([:author, :forum_post_text]).select(select_columns).order((Lms::Forum::ForumPost).TIME + " ASC").limit(limit)
      
    return topic_posts
  end
  
  
  def self.add_reply(id, subject, author_id, text)
    (Lms::Forum::ForumTopic).transaction do
      (Lms::Forum::ForumPost).transaction do
        topic = (Lms::Forum::ForumTopic).find[id]
        raise ActiveRecord::RecordNotFound if topic == nil
        new_post = (Lms::Forum::ForumPost).new(:subject => subject, :author_id => author_id, :forum_topic_id => id)
        new_post.save!
        new_post_text = (Lms::Forum::ForumPostText).new(:text => text, :forum_post_id => new_post[:id])
        new_post_text.save!
        topic[:last_post_id] = new_post[:id]
        return true
      end
    end
  end
  
  
private
  
  def set_time_to_now
    self.time = Time.now
  end
end
